### DISCUSSION TITLE
CIK Lookup - Number

### DISCUSSION THREAD

**Discussion by jdrich79** [sec-edgar/sec-edgar#190] [General]
I was previously able to provide the actual CIK number as the value for 'cik_lookup' and it was working fine. Did something change? Now I get errors, unless I use a ticker symbol. I often want filings for entities without a ticker.

**Comment by jackmoody11** [comment]
> I was previously able to provide the actual CIK number as the value for 'cik_lookup' and it was working fine. Did something change? Now I get errors, unless I use a ticker symbol. I often want filings for entities without a ticker.

What error are you getting? Can you copy and paste it here? We may need to relax the CIK lookup validation.

**Reply by jdrich79** [reply]
I'm certain I was able to do something like this in mid-Feb as I still have the output files and was working on a parser.
```
my_filings = Filing(cik_lookup='0000320193', 
                                filing_type=FilingType.FILING_SC13G, 
                                start_date=datetime(2021, 1, 1)) # end date defaults to today
```

It runs without error, but then the my_filings.save('C:/Users/xyz/filings') I get the following errors:

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~\Anaconda3\lib\site-packages\secedgar\cik_lookup.py in _get_cik_from_html(self, lookup)
    130             span = soup.find('span', {'class': 'companyName'})
--> 131             return span.find('a').getText().split()[0]  # returns single CIK
    132         except AttributeError:  # warn and skip if multiple possibilities for CIK found

AttributeError: 'NoneType' object has no attribute 'find'

During handling of the above exception, another exception occurred:

AttributeError                            Traceback (most recent call last)
~\Anaconda3\lib\site-packages\secedgar\cik_lookup.py in _get_cik_possibilities(soup)
    153             # Exclude table header
--> 154             table_rows = soup.find('table', {'summary': 'Results'}).find_all('tr')[1:]
    155             # Company names are in second column of table

AttributeError: 'NoneType' object has no attribute 'find_all'

During handling of the above exception, another exception occurred:

EDGARQueryError                           Traceback (most recent call last)
<ipython-input-14-2086626f45e5> in <module>
----> 1 my_filings.save('C:/Users/xxxx/SEC_Temp')

~\Anaconda3\lib\site-packages\secedgar\filings\filing.py in save(self, directory, dir_pattern, file_pattern)
    203             ValueError: If no text urls are available for given filing object.
    204         """
--> 205         urls = self._check_urls_exist()
    206 
    207         if dir_pattern is None:

~\Anaconda3\lib\site-packages\secedgar\filings\_base.py in _check_urls_exist(self)
     96             urls (dict): Result of `get_urls` method.
     97         """
---> 98         urls = self.get_urls()
     99         if all(len(urls[cik]) == 0 for cik in urls.keys()):
    100             raise ValueError("No filings available.")

~\Anaconda3\lib\site-packages\secedgar\filings\filing.py in get_urls(self, **kwargs)
    145         return {
    146             key: self._get_urls_for_cik(cik, **kwargs)
--> 147             for key, cik in self.cik_lookup.lookup_dict.items()
    148         }
    149 

~\Anaconda3\lib\site-packages\secedgar\cik_lookup.py in lookup_dict(self)
     86         """:obj:`dict`: Dictionary that makes tickers and company names to CIKs."""
     87         if self._lookup_dict is None:
---> 88             self._lookup_dict = self.get_ciks()
     89         return self._lookup_dict
     90 

~\Anaconda3\lib\site-packages\secedgar\cik_lookup.py in get_ciks(self)
    202             else:
    203                 try:
--> 204                     result = self._get_cik_from_html(lookup)
    205                     self._validate_cik(result)  # raises CIKError if not valid CIK
    206                     ciks[lookup] = result

~\Anaconda3\lib\site-packages\secedgar\cik_lookup.py in _get_cik_from_html(self, lookup)
    133             warning_message = """Lookup '{0}' will be skipped.
    134                           Found multiple companies matching '{0}':
--> 135                           {1}""".format(lookup, '\n'.join(self._get_cik_possibilities(soup)))
    136             warnings.warn(warning_message)
    137         finally:

~\Anaconda3\lib\site-packages\secedgar\cik_lookup.py in _get_cik_possibilities(soup)
    157         except AttributeError:
    158             # If there are no CIK possibilities, then no results were returned
--> 159             raise EDGARQueryError
    160 
    161     @staticmethod

EDGARQueryError:

**Comment by jackmoody11** [comment]
@jdrich79 I did a little debugging and found that this is due to an external change by the SEC. You can see this yourself by trying to go to the following link:

[https://www.sec.gov/cgi-bin/browse-edgar?action=getcompany&CIK=0000320193](https://www.sec.gov/cgi-bin/browse-edgar?action=getcompany&CIK=0000320193).

You will see that you are redirected to 

[https://www.sec.gov/edgar/browse/?CIK=0000320193](https://www.sec.gov/edgar/browse/?CIK=0000320193).

We will need to decide what changes need to be made to the `CIKLookup` class to accommodate this change.

Thanks for pointing this out. I think this highlights the importance of having some sort of separate CI process that checks whether the package works in practice, as we are currently testing using mocks that try to replicate how the website acted at some point in time.

If you would like to help with this, we would appreciate the help and can get that put into the package sooner.

@reteps maybe we should consider adding a raw CIK lookups option that doesn't do any matching and assumes it is given valid CIKs. I am imagining something like this:

```python
# check_lookups would default to True by default but could be set to false to circumvent validation
CompanyFilings(cik_lookup=['0000320193'], filing_type=FilingType.FILING_10K, check_lookups=False)
```

**Reply by reteps** [reply]
@jackmoody11 `re.test(r'\d{8,}')` -> CIK lookup raw

**Reply by jdrich79** [reply]
@jackmoody11 I'd love to help, but my python is not at the level you guys put together.

Looks like the SEC has a new View they are forcing, in the up-left you can see an option to go the "Classic Version". Not sure if you can somehow tap into that to get back the URL you want?

Please let me know if you find a fix, I was all geared up to use your library for something and was so confused when I couldn't get it to work this morning.

**Reply by jackmoody11** [reply]
> @jackmoody11 `re.test(r'\d{8,}')` -> CIK lookup raw

320193 still works via the SEC EDGAR website.

**Reply by jackmoody11** [reply]
> @jackmoody11 I'd love to help, but my python is not at the level you guys put together.
> 
> Looks like the SEC has a new View they are forcing, in the up-left you can see an option to go the "Classic Version". Not sure if you can somehow tap into that to get back the URL you want?
> 
> Please let me know if you find a fix, I was all geared up to use your library for something and was so confused when I couldn't get it to work this morning.

@jdrich79 thanks for pointing that out. That is very helpful. If we remove the action=getcompany parameter from the URL it seems to work fine. Seems like they only redirect when that is set.

**Reply by jdrich79** [reply]
> > @jackmoody11 I'd love to help, but my python is not at the level you guys put together.
> > Looks like the SEC has a new View they are forcing, in the up-left you can see an option to go the "Classic Version". Not sure if you can somehow tap into that to get back the URL you want?
> > Please let me know if you find a fix, I was all geared up to use your library for something and was so confused when I couldn't get it to work this morning.
> 
> @jdrich79 thanks for pointing that out. That is very helpful. If we remove the action=getcompany parameter from the URL it seems to work fine. Seems like they only redirect when that is set.

Sweet! I'd love to contribute, but need to up my Python skills. When I get there I'll see what I can do to help, but I work full-time and am a CS student, so might be a little bit.

**Reply by jackmoody11** [reply]
> > > @jackmoody11 I'd love to help, but my python is not at the level you guys put together.
> > > Looks like the SEC has a new View they are forcing, in the up-left you can see an option to go the "Classic Version". Not sure if you can somehow tap into that to get back the URL you want?
> > > Please let me know if you find a fix, I was all geared up to use your library for something and was so confused when I couldn't get it to work this morning.
> > 
> > 
> > @jdrich79 thanks for pointing that out. That is very helpful. If we remove the action=getcompany parameter from the URL it seems to work fine. Seems like they only redirect when that is set.
> 
> Sweet! I'd love to contribute, but need to up my Python skills. When I get there I'll see what I can do to help, but I work full-time and am a CS student, so might be a little bit.

I understand @jdrich79. I think this should be pretty easy for us to fix. If you would like to contribute in the future, I would be more than happy to help walk you through a pull request. Could even be something as simple as adding to the documentation.

Thanks for helping report this bug! You are helping make this project better for everybody already.

**Comment by meghamu9** [comment]
Hi I've also been using this library and now I am running into the same error. Has this issue been fixed yet?

**Reply by jackmoody11** [reply]
No, this issue has not been fixed yet. We are aware of the cause of the problem but have not had time to fix it yet. You can follow #192 for updates.

**Comment by jackmoody11** [comment]
@meghamu9 can you try `pip install https://github.com/sec-edgar/sec-edgar` and see if that resolves your issue. I think #184 actually fixes this issue.

**Reply by meghamu9** [reply]
Hi, I tried this and it seems that only a couple ciks are working or if I have a list of ciks one downloads and then an error is thrown.

**Reply by jackmoody11** [reply]
Can you try using your browsers user agent?

**Reply by kevinschaul** [reply]
@meghamu9 I was seeing that too (one or two CIKs would work, then an error would be thrown.) We think we've fixed this. Can you try updating to v0.3.3 and then specifying a user agent string, as in [this usage example](https://sec-edgar.github.io/sec-edgar/usage.html#using-a-user-agent)? I am having success using a user agent string of `MY NAME / COMPANY NAME (EMAIL ADDRESS)`.

**Reply by meghamu9** [reply]
Thank you for the details they were helpful ,it works!

**Comment by kevinschaul** [comment]
FYI I'm getting the same error this morning. It's related to rate-limiting.

Here is the text of the page SEC returned after ~6 successful downloads:

> Your Request Originates from an Undeclared Automated Tool
> 
> To allow for equitable access to all users, SEC reserves the right to limit requests originating from undeclared automated tools. Your request has been identified as part of a network of automated tools outside of the acceptable policy and will be managed until action is taken to declare your traffic.
> 
> Please declare your traffic by updating your user agent to include company specific information.
> 
> For best practices on efficiently downloading information from SEC.gov, including the latest EDGAR filings, visit sec.gov/developer . You can also sign up for email updates on the SEC open data program, including best practices that make it more efficient to download data, and SEC.gov enhancements that may impact scripted downloading processes. For more information, contact opendata@sec.gov .
> 
> For more information, please see the SECâs Web Site Privacy and Security Policy . Thank you for your interest in the U.S. Securities and Exchange Commission.

**Reply by jackmoody11** [reply]
@kevinschaul can you try using the user agent your browser uses?

**Reply by kevinschaul** [reply]
Yep! Setting a user agent has been working for me, thanks!
